pkg/cmd/run.go

```diff
@@ -213,113 +207,240 @@ doppler run --mount secrets.json -- cat secrets.json`,
 			}
 		}
 
-		// retrieve secrets
-		dopplerSecrets := fetchSecrets(localConfig, enableCache, fallbackOpts, metadataPath, nameTransformer, dynamicSecretsTTL, format, secretsToInclude)
+		mountOptions := controllers.MountOptions{
+			Enable:   shouldMountFile,
+			Format:   mountFormat,
+			Path:     mountPath,
+			Template: templateBody,
+			MaxReads: maxReads,
+		}
 
-		// The potentially dangerous secret names only are only dangerous when they are set
-		// as environment variables since they have the potential to change the default shell behavior.
-		// When mounting the secrets into a file these are not dangerous
-		if !shouldMountFile {
-			if err := controllers.CheckForDangerousSecretNames(dopplerSecrets); err != nil {
-				utils.LogWarning(err.Error())
-			}
+		watch := cmd.Flags().Changed("watch")
+
+		if watch && fallbackOpts.Exclusive {
+			utils.LogWarning("--watch has no effect when used with --fallback-only")
+			watch = false
 		}
 
-		if len(secretsToInclude) > 0 {
-			noExitOnMissingIncludedSecrets := cmd.Flags().Changed("no-exit-on-missing-only-secrets")
-			missingSecrets := controllers.MissingSecrets(dopplerSecrets, secretsToInclude)
-			if len(missingSecrets) > 0 {
-				err := fmt.Errorf("the following secrets you are trying to include do not exist in your config:\n- %v", strings.Join(missingSecrets, "\n- "))
-				if noExitOnMissingIncludedSecrets {
-					utils.LogWarning(err.Error())
-				} else {
-					utils.HandleError(err)
+		var c *exec.Cmd
+		var cleanupMount func()
+		var err error
+		var lastSecretsFetch time.Time
+		var lastUpdateEvent time.Time
+		// used to ensure we only run one process at a time
+		var processMutex sync.Mutex
+		// used to ensure we only process one event at a time
+		var watchMutex sync.Mutex
+		// this variable has the potential to be racey, but is made safe by our use of the mutex
+		terminatedByWatch := false
+
+		startProcess := func() {
+			// ensure we can fetch the new secrets before restarting the process
+			secrets := controllers.FetchSecrets(localConfig, enableCache, fallbackOpts, metadataPath, nameTransformer, dynamicSecretsTTL, format, secretsToInclude)
+			secretsFetchedAt := time.Now()
+			if secretsFetchedAt.After(lastSecretsFetch) {
+				lastSecretsFetch = secretsFetchedAt
+			}
+
+			controllers.ValidateSecrets(secrets, secretsToInclude, exitOnMissingIncludedSecrets, mountOptions)
+
+			isRestart := c != nil
+			// terminate the old process
+			if isRestart {
+				terminatedByWatch = true
```

Comment:
This mutation of `terminatedByWatch` is not protected by a mutex, should it be?

Reply:
I don't believe a race is possible here given our use of `watchMutex`. `watchMutex` ensures that `startProcess()` cannot be invoked concurrently except on the initial watch event. The initial call to `startProcess()` (before the initial watch event) would run with `c == nil`, so `terminatedByWatch` won't be mutated. Thus I think we're safe here.

In the worst case, a mutation to `terminatedByWatch` would mean us erroneously restarting an exited process rather than the CLI exiting.

Reply:
That makes sense; thanks for the explanation!